### PR TITLE
Automatically build kickstart documentation

### PR DESCRIPTION
**Pull request by atodorov**
This is a WIP on the automated docs generation but you can give it a try and inspect the result by running `make html` inside the docs/ directory. 

The first two commits are required b/c the docs builder assumes the devel handler has the latest versions of all commands, including deprecated ones. This wasn't the case so I had to update the code to fit this assumption.

What's left on the docs side to do is

* Add a Sphinx extension to render the `versionremoved` markup. I'm suprosed Sphinx doesn't have this already. Should be fairly easy to do.
* Add automatic documentation for sections. Should also be easy.
* There are some commands which have been deprecated and no longer in use, e.g. langsupport. At the moment we're skipping these and I'm not sure how to search for them. Need to investigate, otherwise we may end up with incomplete docs.
* Build command docs for RHEL separately due to the way RHEL command classes inherit from earlier ones and how they duplicate some options/implementation with later Fedora counterparts. I need advice on this one, but the simplest way to go is just build a separate section for RHEL latest.
* Add test case for missing or empty descriptions/help - this will fail and I have no idea what text should go into the docs.

@clumens please review and let me know what you think. I will continue with this in the evening.

### CODE REVIEW

**Comment by clumens** [comment]
So far, so good.

**Comment by atodorov** [comment]
Self note: squashed related commits and added documentation for sections.

One new problem found: All sections have version=self.version which is DEVEL in this case. So every description begins with: New in version DEVEL, while options were added in Fedora4 for example. I can probably work around this in the formatter and ignore the `versionadded` markup here.

**Comment by atodorov** [comment]
@clumens I think this is good to go for now. What it does is:

* adds several backward compatibility command definitions and fixes some deprecations (along with a better test for that)
* Add 2 new scripts which iterate over the options parsers and generate the documentation. For commands we take into account the version we're building for (e.g. DEVEL, RHEL7 (or earlier)) and also make sure to document all commands even if not listed in the current handler map. This adds to the documentation some old commands which were used for a release but then silently dropped. Still we want to have documentation for those. Maybe just improve it by properly deprecating the commands.
* Build separately for Fedora and RHEL (currently DEVEL and RHEL7) and include this under 2 different chapters. This is needed because RHEL and Fedora support different commands and options.
* Added small Sphinx extension to parse the `.. versionremoved` markup

If you are happy with how this looks you can `make html` and host the output somewhere for the world to use.

What is still missing is: 

* Add test case for missing or empty descriptions/help - this will fail and I have no idea what text should go into the docs.
* Fix the problem in https://github.com/rhinstaller/pykickstart/pull/100#issuecomment-239295108
* Clean up TODOs added in the code from my previous commits working on automated documentation

I can deal with the last 3 items separately, mainly because of the failing tests.

**Comment by atodorov** [comment]
@clumens - previous remaining items also implemented. There are only few things left:

1) The new test which checks if help descriptions are empty reveals 107 errors. When you run it you will see which classes, etc. I will need texts to fill the data.

2) There are two interesting TODO comments in `method.py` and `bootloader.py`. It looks like that `method.py` does the same as nfs, cdrom, url and I'm not sure what we need it for.

In `bootloader.py` there is the ` --hvArgs` option, which is present only in RHEL5. It wasn't carried over to RHEL6 or later, nor ported to Fedora. It is missing from the docs, but more importantly this behavior is also missing from newer versions of RHEL and Fedora.  With the current way we inherit command classes between releases we're prone to forget some options in an older version and not carry them over to the new one. 

I expect all versions which aren't necessary to be deprecated or removed in the next command implementation, not silently dropped. Can you advice on this?

File: pykickstart/sections.py [hunk 1]
```diff
@@ -126,6 +126,8 @@ def __init__(self, *args, **kwargs):
 class ScriptSection(Section):
     allLines = True
     description = ""
+    epilog = ""
+    title = ""
```
Comment (clumens) [inline pykickstart/sections.py]: These three (description, epilog, and title) should start with an underscore if it's expected they will not be part of the public API that users of pykickstart can do something with.  To me, they very much feel like internal-only things.

**Comment by clumens** [comment]
Prior to some time in 2013, all the installation method commands were implemented in one file - method.py.  In 2013, I broke them out into their own individual files.  method.py stays around because it's needed for backwards compatibility for API users.

hvArgs was only ever supported in RHEL5.  You would have to go back and look at booty in RHEL5 to figure out what it was for.  Just from the name, it seems to be hypervisor related.  It should be missing from everywhere else, as we didn't continue support.

File: docs/gen_sections_docs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+def _format_title(title, underline = '-'):
+    t = title.replace('|', ' or ')
+    return "\n%s\n%s\n" % (t, underline * len(t))
+
+if __name__ == "__main__":
+    chapter = 4
+    handler = returnClassForVersion(DEVEL)
+    for section_class in [PackageSection, PreScriptSection,
+                          PreInstallScriptSection, PostScriptSection,
+                          TracebackScriptSection, OnErrorScriptSection]:
```
Comment (clumens) [inline docs/gen_sections_docs]: Can we grab a list of sections somehow?  I expect I will need to add more in the future, and I will definitely forget to update this list.

Comment (atodorov) [inline docs/gen_sections_docs]: Done, although I'd prefer to see the sections moving under the commands handlers and become versioned as well.

**Comment by atodorov** [comment]
@clumens what is that status of this PR? Do you need anything more on my side b/c I had the impression it was ready to merge?

**Comment by clumens** [comment]
The status is that I wanted to do a new build before merging this in, and builds have been broken for a while due to one of the other very large patch sets I've merged in lately.  Please do "make PYTHON=python2 test" and you'll see the failures.  These need to be cleaned up before anything else can progress.

**Comment by clumens** [comment]
I just merged this and:

ERROR: In `<class 'driverdisk.F14_DriverDisk'>` help can't be blank
ERROR: In `<class 'driverdisk.FC4_DriverDisk'>` help can't be blank
ERROR: In `<class 'driverdisk.FC3_DriverDisk'>` help can't be blank
ERROR: In `<class 'driverdisk.F12_DriverDisk'>` help can't be blank
ERROR: In `<class 'deviceprobe.FC3_DeviceProbe'>` description can't be blank
ERROR: In `<class 'unsupported_hardware.RHEL6_UnsupportedHardware'>` description can't be blank
ERROR: In `<class 'zfcp.F14_ZFCP'>` description can't be blank
ERROR: In `<class 'zfcp.F12_ZFCP'>` description can't be blank
ERROR: In `<class 'zfcp.FC3_ZFCP'>` description can't be blank
ERROR: In `<class 'multipath.FC6_MultiPath'>` description can't be blank
ERROR: In `<class 'multipath.F24_MultiPath'>` description can't be blank
ERROR: In `<class 'realm.F19_Realm'>` description can't be blank
ERROR: In `<class 'interactive.FC3_Interactive'>` description can't be blank
ERROR: In `<class 'interactive.F14_Interactive'>` description can't be blank
ERROR: In `<class 'lilocheck.FC3_LiloCheck'>` description can't be blank
ERROR: In `<class 'upgrade.F11_Upgrade'>` help can't be blank
ERROR: In `<class 'upgrade.F20_Upgrade'>` help can't be blank
ERROR: In `<class 'xconfig.F10_XConfig'>` help can't be blank
ERROR: In `<class 'xconfig.FC3_XConfig'>` help can't be blank
runTest (tests.commands.repo.F21_TestCase) ... ok
ERROR: In `<class 'xconfig.F9_XConfig'>` help can't be blank
ERROR: In `<class 'xconfig.F14_XConfig'>` help can't be blank
ERROR: In `<class 'xconfig.FC6_XConfig'>` help can't be blank
ERROR: In `<class 'method.F19_Method'>` description can't be blank
ERROR: In `<class 'method.F13_Method'>` description can't be blank
ERROR: In `<class 'method.FC3_Method'>` description can't be blank
ERROR: In `<class 'method.F14_Method'>` description can't be blank
ERROR: In `<class 'method.RHEL6_Method'>` description can't be blank
ERROR: In `<class 'method.F18_Method'>` description can't be blank
ERROR: In `<class 'method.FC6_Method'>` description can't be blank
ERROR: In `<class 'mouse.RHEL3_Mouse'>` description can't be blank
ERROR: In `<class 'mouse.FC3_Mouse'>` description can't be blank
ERROR: In `<class 'btrfs.RHEL7_BTRFS'>` help can't be blank
ERROR: In `<class 'btrfs.F17_BTRFS'>` help can't be blank
ERROR: In `<class 'btrfs.F23_BTRFS'>` help can't be blank
runTest (tests.commands.user.FC6_TestCase) ... ok
runTest (tests.commands.user.F8_TestCase) ... ok
ERROR: In `<class 'partition.F11_Partition'>` help can't be blank
ERROR: In `<class 'partition.F23_Partition'>` help can't be blank
ERROR: In `<class 'partition.RHEL7_Partition'>` help can't be blank
ERROR: In `<class 'partition.F20_Partition'>` help can't be blank
ERROR: In `<class 'partition.RHEL6_Partition'>` help can't be blank
ERROR: In `<class 'partition.F14_Partition'>` help can't be blank
ERROR: In `<class 'partition.F12_Partition'>` help can't be blank
ERROR: In `<class 'partition.FC3_Partition'>` help can't be blank
ERROR: In `<class 'partition.F9_Partition'>` help can't be blank
ERROR: In `<class 'partition.F18_Partition'>` help can't be blank
ERROR: In `<class 'partition.F17_Partition'>` help can't be blank
ERROR: In `<class 'partition.FC4_Partition'>` help can't be blank
ERROR: In `<class 'partition.RHEL5_Partition'>` help can't be blank
runTest (tests.commands.user.F12_TestCase) ... ok
ERROR: In `<class 'raid.F19_Raid'>` help can't be blank
ERROR: In `<class 'raid.F23_Raid'>` help can't be blank
ERROR: In `<class 'raid.RHEL7_Raid'>` help can't be blank
ERROR: In `<class 'raid.F9_Raid'>` help can't be blank
ERROR: In `<class 'raid.F18_Raid'>` help can't be blank
ERROR: In `<class 'raid.F12_Raid'>` help can't be blank
ERROR: In `<class 'raid.F14_Raid'>` help can't be blank
ERROR: In `<class 'raid.RHEL6_Raid'>` help can't be blank
ERROR: In `<class 'raid.F15_Raid'>` help can't be blank
ERROR: In `<class 'raid.F20_Raid'>` help can't be blank
ERROR: In `<class 'raid.F13_Raid'>` help can't be blank
ERROR: In `<class 'raid.F25_Raid'>` help can't be blank
ERROR: In `<class 'network.F21_Network'>` help can't be blank
ERROR: In `<class 'network.RHEL7_Network'>` help can't be blank
ERROR: In `<class 'network.F25_Network'>` help can't be blank
runTest (tests.commands.driverdisk.F12_TestCase) ... ok
ERROR: In `<class 'network.F22_Network'>` help can't be blank
ERROR: In `<class 'network.RHEL4_Network'>` help can't be blank
ERROR: In `<class 'network.F24_Network'>` help can't be blank
runTest (tests.commands.deviceprobe.FC3_TestCase) ... ok
runTest (tests.commands.ClassDefinitionTestCase) ... ok
ERROR: In `<class 'dmraid.FC6_DmRaid'>` description can't be blank
ERROR: In `<class 'dmraid.F24_DmRaid'>` description can't be blank
runTest (tests.commands.unsupported_hardware.Unsupported_Hardware_TestCase) ... ok
ERROR: In `<class 'fcoe.F13_Fcoe'>` description can't be blank
ERROR: In `<class 'fcoe.RHEL7_Fcoe'>` description can't be blank
ERROR: In `<class 'fcoe.F12_Fcoe'>` description can't be blank
runTest (tests.commands.driverdisk.FC3_TestCase) ... ok
ERROR: In `<class 'key.RHEL5_Key'>` description can't be blank
runTest (tests.commands.user.F19_TestCase) ... ok
ERROR: In `<class 'bootloader.F14_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.F19_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.F18_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.F8_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.RHEL7_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.F21_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.FC3_Lilo'>` help can't be blank
ERROR: In `<class 'bootloader.F12_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.FC4_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.F17_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.RHEL5_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.F15_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.RHEL6_Bootloader'>` help can't be blank
ERROR: In `<class 'bootloader.FC3_Bootloader'>` help can't be blank
runTest (tests.commands.driverdisk.FC4_TestCase) ... ok
ERROR: In `<class 'langsupport.FC3_LangSupport'>` description can't be blank
ERROR: In `<class 'langsupport.FC5_LangSupport'>` description can't be blank
runTest (tests.commands.unsupported_hardware.RHEL6_TestCase) ... ok
ERROR: In `<class 'device.F24_Device'>` description can't be blank
ERROR: In `<class 'device.F8_Device'>` description can't be blank
ERROR: In `<class 'device.FC3_Device'>` description can't be blank
runTest (tests.commands.eula.Eula_TestCase) ... ok
runTest (tests.commands.upgrade.F20_TestCase) ... ok
runTest (tests.commands.install.F20_TestCase) ... ok
ERROR: In `<class 'firewall.FC3_Firewall'>` help can't be blank
ERROR: In `<class 'firewall.F10_Firewall'>` help can't be blank
ERROR: In `<class 'firewall.F9_Firewall'>` help can't be blank
ERROR: In `<class 'firewall.F20_Firewall'>` help can't be blank
ERROR: In `<class 'firewall.F14_Firewall'>` help can't be blank
runTest (tests.commands.upgrade.FC3_TestCase) ... ok
ERROR: In `<class 'logvol.F20_LogVol'>` help can't be blank
ERROR: In `<class 'logvol.F14_LogVol'>` help can't be blank
ERROR: In `<class 'logvol.F21_LogVol'>` help can't be blank
ERROR: In `<class 'logvol.RHEL6_LogVol'>` help can't be blank
ERROR: In `<class 'logvol.F9_LogVol'>` help can't be blank
ERROR: In `<class 'logvol.F15_LogVol'>` help can't be blank
ERROR: In `<class 'logvol.F18_LogVol'>` help can't be blank
ERROR: In `<class 'logvol.RHEL7_LogVol'>` help can't be blank
ERROR: In `<class 'logvol.F17_LogVol'>` help can't be blank
ERROR: In `<class 'logvol.F12_LogVol'>` help can't be blank
ERROR: In `<class 'logvol.F23_LogVol'>` help can't be blank
runTest (tests.commands.driverdisk.F14_TestCase) ... ok
runTest (tests.commands.HelpAndDescription_TestCase) ... FAIL
======================================================================
FAIL: runTest (tests.commands.HelpAndDescription_TestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.5/unittest/mock.py", line 1157, in patched
    return func(*args, **keywargs)
  File "/home/clumens/src/pykickstart/tests/commands/__init__.py", line 132, in runTest
    self.assertEqual(0, errors)
AssertionError: 0 != 106

**Comment by atodorov** [comment]
@clumens yeah, see https://github.com/rhinstaller/pykickstart/pull/100#issuecomment-240098229 and let me know how do you prefer to deal with that. Ideally we fix all the missing help documentation worst case you revert the test that generates these errors.

**Comment by M4rtinK** [comment]
Um, what's the plan with online kickstart documentation now ? We point people to the following file [in the wiki](https://github.com/rhinstaller/pykickstart/blob/master/docs/kickstart-docs.rst):
https://github.com/rhinstaller/pykickstart/blob/master/docs/kickstart-docs.rst

But it now kinda lacks any useful content. Also there might possibly be some dead links now that used to point to sections in that file.

So will the generated documentation be soon online somewhere else & automatically kept up to date ? Otherwise I unfortunately see this as a sizeable regression form the documentation point of view.

**Comment by clumens** [comment]
Martin - I suppose I'll upload the documentation every time I do a build of the package.

**Comment by atodorov** [comment]
> Um, what's the plan with online kickstart documentation now ? We point people to the following file in the wiki:
https://github.com/rhinstaller/pykickstart/blob/master/docs/kickstart-docs.rst

we can update the wiki

> But it now kinda lacks any useful content. Also there might possibly be some dead links now that used to point to sections in that file.

Maybe we can add a link to the new location in this file so we at least keep the URL compatibility. Page anchors (e.g. #links) should still work.

@clumens - I think we're also ReadTheDocs compatible if not ready. Is that an option for the documentation ?

**Comment by clumens** [comment]
@atodorov - On the previous comment - it looks like the only thing left to decide is whether we're going to add docs for those 100-some missing commands?  Could we turn the test that is current failing into something that is just informational while we add what's missing, and then make it a blocker once that's done so nothing gets skipped in the future?

**Comment by clumens** [comment]
@atodorov - I'm fine with readthedocs, as long as the wiki still points in the right place and the built package contains docs.

**Comment by M4rtinK** [comment]
> @atodorov - I'm fine with readthedocs, as long as the wiki still points in the right place and the built package contains docs.

Yeah, if we could get it working then I also think readthedocs is the best option. People generally expect project documentation to be there these days & it looks like it can automatically update the generated documentation if the repo changes, so that would be one less thing to handle by ourselves. :)
